packages/apollo-server-core/src/plugin/usageReporting/referencedFields.ts

```diff
@@ -0,0 +1,82 @@
+import {
+  DocumentNode,
+  GraphQLSchema,
+  isInterfaceType,
+  separateOperations,
+  TypeInfo,
+  visit,
+  visitWithTypeInfo,
+} from 'graphql';
+import { ReferencedFieldsForType } from 'apollo-reporting-protobuf';
+
+export type ReferencedFieldsByType = Record<string, ReferencedFieldsForType>;
+
+export function calculateReferencedFieldsByType({
+  document,
+  schema,
+  resolvedOperationName,
+}: {
+  document: DocumentNode;
+  resolvedOperationName: string | null;
+  schema: GraphQLSchema;
+}): ReferencedFieldsByType {
+  // If the document contains multiple operations, we only care about fields
+  // referenced in the operation we're using and in fragments that are
+  // (transitively) spread by that operation. (This is because Studio's field
+  // usage accounting is all by operation, not by document.) This does mean that
+  // a field can be textually present in a GraphQL document (and need to exist
+  // for validation) without being represented in the reported referenced fields
+  // structure, but we'd need to change the data model of Studio to be based on
+  // documents rather than fields if we wanted to improve that.
+  const documentSeparatedByOperation = separateOperations(document);
+  const filteredDocument =
+    documentSeparatedByOperation[resolvedOperationName ?? ''];
+  if (!filteredDocument) {
+    // This shouldn't happen because we only should call this function on
+    // properly executable documents.
+    throw Error(
+      `shouldn't happen: operation '${resolvedOperationName ?? ''}' not found`,
+    );
+  }
+  const typeInfo = new TypeInfo(schema);
+  const interfaces = new Set<string>();
+  const referencedFieldSetByType: Record<string, Set<string>> = Object.create(
+    null,
+  );
+  visit(
+    filteredDocument,
+    visitWithTypeInfo(typeInfo, {
+      Field(field) {
+        const fieldName = field.name.value;
+        const parentType = typeInfo.getParentType();
+        if (!parentType) {
+          throw Error(
+            `shouldn't happen: missing parent type for field ${fieldName}`,
+          );
+        }
+        const parentTypeName = parentType.name;
+        if (!referencedFieldSetByType[parentTypeName]) {
+          referencedFieldSetByType[parentTypeName] = new Set<string>();
+          if (isInterfaceType(parentType)) {
+            interfaces.add(parentTypeName);
+          }
+        }
+        referencedFieldSetByType[parentTypeName].add(fieldName);
```

Comment:
Feels like the `mapGetOrSet` pattern we've been using lately. Is there a reason you chose a `Record` for `referencedFieldSetByType` instead?

Reply:
Because there's no mapGetOrSet in apollo-server yet :)

Note that for the actual protobuf type it does need to be a record, but this could be a Map since I'm immediately converting it to the real type.

On the other hand, the downside to mapGetOrSet is that it constructs the "default" on every call even if it's not used.  So usually that's making a single garbage array.  Here that would be creating a garbage Set (idk if that's more overhead than an array) and also it's not clear where you'd put the `isInterfaceType` call.  Since this does have the isInterface thing, maybe I'll leave this as it is... but also not be hesitant to add mapGetOrSet next time? What do you think?

Reply:
SGTM!